### ISSUE TITLE
 lock document with signature

### ISSUE THREAD

**Issue by fredtucs** [LibrePDF/OpenPDF#479]
Is there a way to block the document or textfield with digital signature?

**Comment by Lonzak** [comment]
When you digitally sign a document it is "blocked" so to speak. Conforming readers should not allow to change it. There are tools however (like the apple preview) which completely ignores this and allow changes to the document and thus break it. 
What we additionally do before signing the first signature, is that we set the readonly and lock flags for all form fields. (But even that is no guarantee that for some tools...)
You can not prevent, that somebody changes/manipulates/corrupts the document. However the good thing is that you are able to verify the integrity of the file and thus detect all changes...

**Comment by fredtucs** [comment]
Thanks for your answer @Lonzak 

Perhaps I have not been very precise in my query, the case that I am trying to solve is that several signatures are inserted in the document and before the last signature the fields are filled in is where I need to block the fields.

Reviewing I could verify that it is possible to do it with iText 5.5.6 by means of the following code:

```
PdfFormField field = PdfFormField.createSignature(stamper.getWriter());
field.setFieldName("Signature");
field.put(PdfName.LOCK, stamper.getWriter().addToBody(new PdfSigLockDictionary(LockPermissions.NO_CHANGES_ALLOWED)).getIndirectReference());
field.setFlags(PdfAnnotation.FLAGS_PRINT);
field.setPage(1);
field.setWidget(new Rectangle(150, 250, 300, 401), PdfAnnotation.HIGHLIGHT_INVERT);
stamper.addAnnotation(field, 1);
```

**however I would like to know if it is also possible to do it with OpenPDF.**

thank you for your help.

**Comment by Lonzak** [comment]
Yes that is quite possible. However note, that applying a LOCK to a field doesn't mean that filling the field is blocked.
The LOCK flag means that the properties can not be changed.  For that you need to apply the READ-ONLY flag. (which can also be applied by openPDF)

The spec says about the LOCK flag:
"If set, do not allow the annotation to be **deleted** or its **properties** (including position and size) to be modified by the user. However, this flag does **not restrict changes to the annotation’s contents**, such as the value of a form field."

**Comment by mkl-public** [comment]
The OP does not talk about the lock _flag_, he talks about _signature field lock dictionaries_.

**Comment by Lonzak** [comment]
Ah you are right, was too skimming to quickly :-)
But still, adding a LOCK dictionary with openPDF is also possible...

**Comment by mkl-public** [comment]
Indeed, the `PdfSigLockDictionary` class is no magic, merely a dictionary with a very few entries.
But one should be aware that there is no explicit OpenPDF support for the lock dictionary during signing; in iText such support started during the 5.3.x versions. This might be of interest if the OP wants to apply that "last signature" using OpenPDF. (It might work out of the box, though. If I recall correctly, the feature the OP wants to use doesn't require further actions during signing.)

**Comment by fredtucs** [comment]
First of all thank you for your answers.

I am just trying to do what @mkl-public mentions in your comment, maybe you could help me to better understand the idea by using a code snippet how to add a LOCK to the dictionary with openPDF, similar to what the `PdfSigLockDictionary` does.

thank you for your help.

**Comment by asturio** [comment]
Improvements in pull requests are welcome!
